GPSClick.device.nut

```diff
@@ -0,0 +1,233 @@
+                break;
+            }
+            check = check ^ i;
+        }
+        return check;
+    }
+    
+    function parse_fields() {
+        local str = "";
+        my_fields = [];
+        foreach(i in p_sentence) {
+            if(i == ',') {
+                my_fields.push(str);
+                str = "";
+            }
+            else {
+                str+=i.tochar();
+            }
+        }
+        if(str.len() > 0) {
+            my_fields.push(str);
+        }
+    }
+    
+    function printFields() {
+        if(my_fields != null && my_fields.len() > 0) {
+            server.log(my_fields[0]);
+    }
+    }
+    
+    function extractData() {
+        local retTable = {};
+        if(my_fields != null && my_fields.len() > 0) {
+            switch(my_fields[0]) {
+                case VTG:
+                    retTable.type <- VTG;
+                    if(my_fields[1] != "") {
+                        retTable.trackt <- my_fields[1];
+                    }
+                    if(my_fields[7] != "") {
+                        retTable.speedkmh <- my_fields[7];
+                    }
+                    
+                    local checkLen = my_fields[my_fields.len() - 1].len();
+                    retTable.checkSum <- my_fields[my_fields.len()-1].slice(checkLen-4, checkLen-2);
+                    break;
+
+                case RMC:
+                    retTable.type <- RMC;
+                    local lat = my_fields[3];
+                    if(lat.len() <= 1) break; // no data
+                    local time = my_fields[1].tointeger();
+                    
+                    retTable.seconds <- time%100;
+                    time = time/100;
+                    retTable.minutes <- time%100;
+                    time = time/100;
+                    retTable.hours <- time; 
+                    
+                    retTable.status <- (my_fields[2] == "A" ? "Active" : "Void");
+                    
+                    local latitude = lat.slice(0, 2).tofloat() + 
+                    lat.slice(2).tofloat()/60;
+                    if(my_fields[4] == "S") latitude = -latitude;
+                    retTable.latitude <- latitude;
+                    
+                    local long = my_fields[5];
+                    local longitude = long.slice(0, 3).tofloat() + 
+                    long.slice(3).tofloat()/60;
+                    if(my_fields[6] == "W") longitude = -longitude;
+                    retTable.longitude <- longitude;
+                    
+                    local checkLen = my_fields[my_fields.len() - 1].len();
+                    retTable.checkSum <- my_fields[my_fields.len()-1].slice(checkLen-4, checkLen-2);
+                    break;
+                case GLL:
+                    retTable.type <- GLL;
+                    local lat = my_fields[1];
+                    if(lat.len() <= 1) break;
+                    
+                    local latitude = lat.slice(0, 2).tofloat() +
```

Comment:
A number of the calculations in these cases are quite similar.  Look at ways to refactor so you don't have to keep writing the same thing over and over.